packager/media/formats/mp2t/es_parser_teletext.cc

```diff
@@ -0,0 +1,518 @@
+      UpdateNationalSubset(TELETEXT_NATIONAL_SUBSET_ITALIAN);
+      break;
+    case TELETEXT_NATIONAL_SUBSET::NONE:
+    default:
+      break;
+  }
+}
+
+void EsParserTeletext::SendPending(const uint16_t index, const int64_t pts) {
+  auto page_state_itr = page_state_.find(index);
+
+  if (page_state_itr == page_state_.end() ||
+      page_state_itr->second.lines.empty()) {
+    return;
+  }
+
+  const auto& pending_lines = page_state_itr->second.lines;
+  const auto pending_pts = page_state_itr->second.pts;
+
+  TextFragmentStyle text_fragment_style;
+  TextSettings text_settings;
+  std::shared_ptr<TextSample> text_sample;
+
+  if (pending_lines.size() == 1) {
+    TextFragment text_fragment(text_fragment_style, pending_lines[0].c_str());
+    text_sample = std::make_shared<TextSample>("", pending_pts, pts,
+                                               text_settings, text_fragment);
+
+  } else {
+    std::vector<TextFragment> sub_fragments;
+    for (const auto& line : pending_lines) {
+      sub_fragments.emplace_back(text_fragment_style, line.c_str());
+      sub_fragments.emplace_back(text_fragment_style, true);
+    }
+    sub_fragments.pop_back();
+    TextFragment text_fragment(text_fragment_style, sub_fragments);
+    text_sample = std::make_shared<TextSample>("", pending_pts, pts,
+                                               text_settings, text_fragment);
+  }
+
+  text_sample->set_sub_stream_index(index);
+  emit_sample_cb_(text_sample);
+
+  page_state_.erase(index);
+}
+
+std::string EsParserTeletext::BuildText(const uint8_t* data_block,
+                                        const uint8_t row) const {
+  const size_t payload_len = 40;
+  std::string next_string;
+  next_string.reserve(payload_len * 2);
+  bool leading_spaces = true;
+
+  const uint16_t index = magazine_ * 100 + page_number_;
+  const auto page_state_itr = page_state_.find(index);
+
+  const std::unordered_map<uint8_t, std::string>* column_replacement_map =
+      nullptr;
+  if (page_state_itr != page_state_.cend()) {
+    const auto row_itr =
+        page_state_itr->second.packet_26_replacements.find(row);
+    if (row_itr != page_state_itr->second.packet_26_replacements.cend()) {
+      column_replacement_map = &(row_itr->second);
+    }
+  }
+
+  for (size_t i = 0; i < payload_len; ++i) {
+    if (column_replacement_map) {
+      const auto column_itr = column_replacement_map->find(i);
+      if (column_itr != column_replacement_map->cend()) {
+        next_string.append(column_itr->second);
+        leading_spaces = false;
+        continue;
+      }
+    }
+
+    char next_char =
+        static_cast<char>(TELETEXT_BITREVERSE_8[data_block[i]] & 0x7f);
+
+    if (next_char < 32) {
+      next_char = 0x20;
+    }
+
+    if (leading_spaces) {
+      if (next_char == 0x20) {
+        continue;
+      }
+      leading_spaces = false;
+    }
+
+    switch (next_char) {
+      case '&':
+        next_string.append("&amp;");
```

Comment:
I'm not very familiar with this part of the code.  Why does the text need to be XML-escaped?